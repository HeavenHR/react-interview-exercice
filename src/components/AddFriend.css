:local(.addFriendContainer) {
  margin: 10px;
}

:local(.addFriendButtonHolder) {
  margin-bottom: 10px;
}

:local(.tickbox) {
  margin: 10px; 
  display: inline;
}

:local(.color) {
  color: lightgreen;
}
