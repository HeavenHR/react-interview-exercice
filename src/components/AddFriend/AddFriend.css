:local(.addFriend) {
  background: white;
  border-radius: 4px;
  padding: 16px;
  margin-bottom: 16px;
}