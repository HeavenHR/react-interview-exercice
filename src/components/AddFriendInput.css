:local(.addFriendInput) {
  border-radius: 0;
  border-color: #abaaaa;
  border-left: 0;
  border-right: 0;
}

:local(.formContainer) {
  height: 40px;
}

:local(.genderDropdown) {
  height: 30px;
  width: 160px;
  margin: 5px 5px;
}

:local(.btnSubmit) {
  height: 30px;
  border-radius: 5px;
  margin-left: 15px;
}
