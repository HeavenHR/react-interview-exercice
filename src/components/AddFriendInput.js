import React, { Component } from "react";
import { PropTypes } from "prop-types";
import classnames from "classnames";
import styles from "./AddFriendInput.css";

class AddFriendInput extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      name: this.props.name || "",
      gender: ""
    };
  }

  handleChange(e) {
    this.setState({ name: e.target.value });
  }

  handleSelectChange(e) {
    this.setState({ gender: e.target.value });
  }

  handleSubmit(e) {
    e.preventDefault();
    const name = e.target[0].value.trim();
    const gender = e.target[1].value;
    this.props.addFriend(name, gender);
    this.setState({ name: "", gender: "" });
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit.bind(this)}>
        <input
          type="text"
          autoFocus="true"
          className={classnames("form-control", styles.addFriendInput)}
          placeholder="Type the name of a friend"
          value={this.state.name}
          onChange={this.handleChange.bind(this)}
        />
        <div className={styles.formContainer}>
          <select
            name="gender"
            value={this.state.gender}
            onChange={this.handleSelectChange.bind(this)}
            className={styles.genderDropdown}
          >
            <option value="">select gender</option>
            <option value="Male">Male</option>
            <option value="Female">Female</option>
          </select>
          <input
            type="submit"
            value="Add Friend"
            className={styles.btnSubmit}
          />
        </div>
      </form>
    );
  }
}

AddFriendInput.propTypes = {
  addFriend: PropTypes.func.isRequired
};

export default AddFriendInput;
