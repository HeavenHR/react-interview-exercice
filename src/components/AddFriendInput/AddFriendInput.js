import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import styles from './AddFriendInput.css';
import ErrorMessage from '../ErrorMessage';

class AddFriendInput extends Component {
  static propTypes = {
    value: PropTypes.string,
    onSubmit: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    isInvalid: PropTypes.bool,
    errorMessage: PropTypes.string
  }

  render() {
    const { isInvalid, errorMessage } = this.props;
    return (
      <div className={classnames('form-group', { 'has-error': isInvalid })}>
        <input
          type="text"
          autoFocus="true"
          className={classnames('form-control', styles.addFriendInput)}
          placeholder="Type the name of a friend"
          value={this.state.value}
          onChange={this.handleChange.bind(this)}
          onKeyDown={this.handleSubmit.bind(this)}
        />
        {isInvalid && errorMessage && <ErrorMessage message={errorMessage} />}
      </div>
    );
  }

  constructor(props, context) {
    super(props, context);
    this.state = {
      value: this.props.value || ''
    };
  }

  componentWillReceiveProps(nextProps) {
    const { value } = nextProps;

    if (value !== this.props.value) {
      this.setState({ value });
    }
  }

  handleChange(e) {
    const value = e.target.value;
    const { onChange } = this.props;

    this.setState({ value });
    onChange && onChange(value);
  }

  handleSubmit(e) {
    const { onSubmit } = this.props;
    if (e.which === 13) {
      onSubmit && onSubmit();
    }
  }
}

export default AddFriendInput
