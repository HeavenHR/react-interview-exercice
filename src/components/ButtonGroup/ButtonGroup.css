.btn {
  text-transform: capitalize;
}

.btn-primary:active {
  background-color: #337ab7 !important;
  border-color: #2e6da4 !important;
}

.fa-male, .fa-female {
  padding-right: 10px;
}

.form-group:last-child {
  margin-bottom: 0;
}