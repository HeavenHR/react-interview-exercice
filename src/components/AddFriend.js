import React, { Component, PropTypes } from "react";
import classnames from "classnames";

import AddFriendInput from "./AddFriendInput";
import AddFriendRadio from "./AddFriendRadio";

import { MALE } from "./../constants/PageTypes";

import styles from "./AddFriend.css";

class AddFriend extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      name: "",
      gender: MALE
    };
  }

  render() {
    return (
      <div onKeyDown={this.handleKeyDown} className={styles.addFriendContainer}>
        <AddFriendInput name={this.state.name} onHandleChange={this.handleChange} />
        <AddFriendRadio gender={this.state.gender} onHandleChange={this.handleChange} />
        <div className={classnames("text-center", styles.addFriendButtonHolder)}>
          <button
            type='button'
            className='btn'
            onClick={this.handleSubmit}
            disabled={this.state.name.length > 0 ? false : true}>
            Add Friend
          </button>
        </div>
      </div>
    );
  }

  resetState = () => {
    this.setState({ name: "", gender: MALE });
  };

  handleChange = (key, value) => {
    this.setState({ [key]: value });
  };

  handleSubmit = () => {
    const nameFromState = this.state.name.trim();
    this.props.addFriend({ name: nameFromState, gender: this.state.gender });
    this.resetState();
  };

  handleKeyDown = (e) => {
    const nameFromState = this.state.name.trim();
    if (e.which === 13 && nameFromState.length > 0) {
      this.handleSubmit();
    }
  };
}

AddFriend.propTypes = {
  addFriend: PropTypes.func.isRequired
};

export default AddFriend;
