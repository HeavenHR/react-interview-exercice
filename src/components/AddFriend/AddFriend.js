import React, { Component } from 'react';
import PropTypes from 'prop-types';
import className from 'classnames/bind';
import omit from 'lodash/omit';

import AddFriendInput from '../AddFriendInput';
import ButtonsGroup from '../ButtonGroup';

import styles from './AddFriend.css';

const cn = className.bind(styles);

class AddFriend extends Component {
  static propTypes = {
    addFriend: PropTypes.func.isRequired
  }

  state = {
    name: '',
    gender: '',
    validation: {}
  }

  validation = (newGender) => {
    const { name, gender } = this.state;
    const updatedGender = newGender || gender;

    const validation = {
      nameInvalid: !name,
      genderInvalid: !updatedGender
    };

    this.setState({ validation });

    return !Object.values(validation).includes(true);
  }

  resetState = () => this.setState({ name: '', gender: '', validation: {} });

  handleGenderSelect = gender => {
    const { addFriend } = this.props;

    if (this.validation(gender)) {
      addFriend && addFriend({ ...omit(this.state, 'validation'), gender });
      this.resetState();
    } else {
      this.setState({ gender });
    }
  }

  handleNameChange = name => {
    this.setState({ name, validation: {} });
  }

  handleSubmit = () => {
    const { addFriend } = this.props;

    if (this.validation()) {
      addFriend && addFriend(omit(this.state, 'validation'));
      this.resetState();
    }
  }

  options = [
    {value: 'male', showIcon: true},
    {value: 'female', showIcon: true}
  ]

  render() {
    const { name, gender, validation: { nameInvalid, genderInvalid } } = this.state;

    return (
      <div className={cn('addFriend')}>
        <AddFriendInput
          value={name}
          onSubmit={this.handleSubmit}
          onChange={this.handleNameChange}
          isInvalid={nameInvalid}
          errorMessage="Please fill friend name"
        />
        <ButtonsGroup
          selected={gender}
          options={this.options}
          onClick={this.handleGenderSelect}
          isInvalid={genderInvalid}
          errorMessage="Please select gender"
        />
      </div>
    );
  }
}

export default AddFriend;
