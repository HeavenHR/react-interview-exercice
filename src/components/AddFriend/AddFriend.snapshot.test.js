import React from 'react';

import AddFriend from './AddFriend';
import AddFriendInput from '../AddFriendInput';
import ButtonGroup from '../ButtonGroup';

import { createMountedComponent, sampleFriend } from '../../utils/testutil';


describe('<AddFriend /> should', () => {
  const mountedAddFriend = createMountedComponent({
    children: <AddFriend addFriend={localStorage.actions.addFriend} />
  });

  const inputField = mountedAddFriend.find(AddFriendInput);
  const buttonGroup = mountedAddFriend.find(ButtonGroup);

  it('render correctly', () => {
    expect(mountedAddFriend.htmlRender()).toMatchSnapshot();
    expect(inputField).toHaveLength(1);
    expect(buttonGroup).toHaveLength(1);
  });

  it('simulate addFriend action when click gender button with friend name', () => {
    inputField.props().onChange(sampleFriend.name);
    expect(mountedAddFriend.state().name).toEqual(sampleFriend.name);
    buttonGroup.props().onClick(sampleFriend.gender);
    expect(mountedAddFriend.props().addFriend).toBeCalledWith(sampleFriend);
    // console.log(mountedAddFriend.state());
    expect(mountedAddFriend.state()).toEqual({ name: '', gender: '', validation: {} });
  });

  it('validate name and gender on input submit', () => {
    inputField.props().onSubmit();
    expect(mountedAddFriend.state().validation).toEqual({ 'genderInvalid': true, 'nameInvalid': true });
  })

  it('validate name on gender selection', () => {
    buttonGroup.props().onClick(sampleFriend.gender);
    expect(mountedAddFriend.state().validation).toEqual({ 'genderInvalid': false, 'nameInvalid': true });
  })
});
