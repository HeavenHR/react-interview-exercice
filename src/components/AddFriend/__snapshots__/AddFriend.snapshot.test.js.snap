exports[`<AddFriend /> should render correctly 1`] = `
"<div class=\"addFriend\">
  <div class=\"form-group\">
    <input type=\"text\" class=\"form-control\" placeholder=\"Type the name of a friend\"
    value=\"\">
  </div>
  <div class=\"form-group\">
    <div class=\"btn-group btn-block\">
      <button type=\"button\" value=\"male\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-male\"></i>
        <!-- react-text: 8 -->male
        <!-- /react-text -->
      </button>
      <button type=\"button\" value=\"female\" class=\"btn btn-primary col-sm-6\"><i class=\"fa fa-female\"></i>
        <!-- react-text: 11 -->female
        <!-- /react-text -->
      </button>
    </div>
  </div>
</div>"
`;
